docs/rfcs/038-independent-compute-release.md

```diff
@@ -0,0 +1,298 @@
+
+### Compute config/settings manifest
+
+We will create a new sub-directory `compute` and file `compute/manifest.yaml` with a structure:
+
+```yaml
+pg_settings:
+  # Common settings for primaries and secondaries of all versions.
+  common:
+    wal_log_hints: "off"
+    max_wal_size: "1024"
+
+  per_version:
+    14:
+      # Common settings for both replica and primary of version PG 14
+      common:
+        shared_preload_libraries: "neon,pg_stat_statements,extension_x"
+    15:
+      common:
+        shared_preload_libraries: "neon,pg_stat_statements,extension_x"
+      # Settings that should be applied only to
+      replica:
+        # Available only starting Postgres 15th
+        recovery_prefetch: "off"
+    # ...
+    17:
+      common:
+        # For example, if third-party `extension_x` is not yet available for PG 17
+        shared_preload_libraries: "neon,pg_stat_statements"
+      replica:
+        recovery_prefetch: "off"
+```
+
+N.B.: setting value should be a string with `on|off` for booleans and a number (as a string)
+without units for all numeric settings. That's how control plane currently operates.
+
+The priority of settings will be (higher number is a higher priority):
+
+1. Any static and hard-coded settings in the control plane
+2. `pg_settings->common`
+3. Per-version `common`
+4. Per-version `replica`
+5. Any per-user/project/endpoint overrides in the control plane
+6. Any dynamic setting calculated based on the compute size
+
+Compute team will own the content of `compute/manifest.yaml`.
+
+### Control plane: releases table
+
+In order to store information about releases, control plane will use a table `compute_releases` with the following
+schema:
+
+```sql
+CREATE TABLE compute_releases (
+  -- Unique release ID
+  revision BIGSERIAL PRIMARY KEY,
+  -- Numeric version of the compute image, e.g. 5432
+  version BIGINT NOT NULL,
+  -- Compute image tag, e.g. `release-5432`
+  tag TEXT NOT NULL,
+  -- Current release status. Currently, it will be a simple enum
+  -- * `deployed` -- release is deployed and used for new compute starts.
+  --                 Exactly one release can have this status at a time.
+  -- * `superseded` -- release has been replaced by a newer one.
+  -- But we can always extend it in the future, when we will need more statuses
+  -- for more complex deployment strategies.
+  status TEXT NOT NULL,
+  -- Any additional metadata for compute in the corresponding release
+  manifest JSONB NOT NULL,
+  -- Timestamp when release record was created in the control plane database
+  created_at TIMESTAMP NOT NULL DEFAULT now(),
+  -- Timestamp when release deployment was finished
+  deployed_at TIMESTAMP
+);
+```
+
+### Control plane: management API
+
+Control plane will implement new API methods to manage releases:
+
+1. `POST /management/api/v2/compute_releases` to create a new release. With payload
+
+   ```json
+    {
+      "version": 5432,
+      "tag": "release-5432",
+      "manifest": {}
```

Comment:
I assume the `manifest` will come from the `compute/manifests.yaml` file.

That makes the API a bit more complicated to use this API manually. Could we include the settings directly in the compute images?

Reply:
I think we can override what's in the manifests in the cplane so that's why we want this information to be available to cplane.

Though, I agree this is going to make manual use of this API quite a pain :p

Reply:
Yeah, as per the overrides priority, cplane can have a higher level overrides. That means we would need to extract the manifest from the Docker image and put into the cplane again, which is even worse. We can still consider bundling it into a Docker image because it's cheap, and one could always copy it out of it

As for the manual API call, I'd expect us to never do it manually. Only in case GitHub is down. In all other cases we should release via triggering a CI job, same as we do for all other services. I do agree that we may need to do it completely manually, so we will just prepare a one-liner with `cat ... | yq ...` to convert it to json and put into `curl`